### PR TITLE
fix(v2): keep DocSearch state on remounts

### PR DESCRIPTION
**Pull request by francoischalifour**
## Motivation

The Docusaurus navbar remounts at a certain screen resize. This creates new variable references for objects and functions passed to the `DocSearchModal` component, resulting in a rerender of DocSearch that then loses the previous state (because we listen to these reference changes in our component).

This PR keeps references of DocSearch options by using `useMemo` and `useRef`.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/facebook/docusaurus/blob/master/CONTRIBUTING.md#pull-requests)?

Yes.

### Before

docusaurus-ds3-before

### After

docusaurus-ds2-fix-after

### CODE REVIEW

File: packages/docusaurus-theme-search-algolia/src/theme/SearchBar/index.js [hunk 1]
```diff
@@ -73,6 +73,38 @@ function DocSearch(props) {
...
+  const resultsFooterComponent = useMemo(
+    () => (footerProps) => <ResultsFooter {...footerProps} onClose={onClose} />,
+    [onClose],
+  );
+
+  const transformSearchClient = useRef((searchClient) => {
+    searchClient.addAlgoliaAgent('docusaurus', siteMetadata.docusaurusVersion);
+
+    return searchClient;
+  }).current;
```
Comment (slorber) [inline packages/docusaurus-theme-search-algolia/src/theme/SearchBar/index.js:106]: oooh this is a funny pattern I've never seen before :)

**Comment by francoischalifour** [comment]
@slorber `history` and `withBaseUrl`'s references change at every render so I cannot use `useMemo` and `useCallback` here. Do you have a solution to avoid `useRef` here?

(I was able to use `useCallback` for `transformSearchClient`.)

**Comment by slorber** [comment]
Ahah I see it's this annoying problem again :) 
Let's merge this and move on

(see https://github.com/facebook/react/issues/16956)

**Comment by slorber** [comment]
Ahah I see it's this annoying problem again :) 
Let's merge this and move on

(see https://github.com/facebook/react/issues/16956)

**Comment by slorber** [comment]
one suggestion culd be to handle the memoization directly in the lib, and wrap your component in a parent component whose role is to make the props "stable"

That's a bit annoying but at least it could prevent end users to have to do this themselves, so the lib is more idiomatic to use

**Pull request merged**